packages/bus-query-operation/lib/Bindings.ts

```diff
@@ -50,3 +52,195 @@ export function isBindings(maybeBindings: any): boolean {
+          ),
+        };
+      }
+    },
+    project: (op: Algebra.Project, factory: Factory) => {
+      // Materialize a project operation.
+      // If strictTargetVariables is true, we throw if the project target variable is attempted to be bound.
+      // Otherwise, we just filter out the bound variables.
+      if (strictTargetVariables) {
+        for (const variable of op.variables) {
+          if (bindings.has(termToString(variable))) {
+            throw new Error(`Tried to bind variable ${termToString(variable)} in a SELECT operator.`);
+          }
+        }
+        return {
+          recurse: true,
+          result: op,
+        };
+      } else {
+        const variables = op.variables.filter((variable) => !bindings.has(termToString(variable)));
+        return {
+          recurse: true,
+          result: factory.createProject(
+            op.input,
+            variables,
+          ),
+        };
+      }
+    },
+    values: (op: Algebra.Values, factory: Factory) => {
+      // Materialize a values operation.
+      // If strictTargetVariables is true, we throw if the values target variable is attempted to be bound.
+      // Otherwise, we just filter out the bound variables and their bindings.
+      if (strictTargetVariables) {
+        for (const variable of op.variables) {
+          if (bindings.has(termToString(variable))) {
+            throw new Error(`Tried to bind variable ${termToString(variable)} in a VALUES operator.`);
+          }
+        }
+      } else {
+        const variables = op.variables.filter((variable) => !bindings.has(termToString(variable)));
+        const valueBindings = op.bindings.map((binding) => {
+          const newBinding = { ...binding };
+          bindings.forEach((value, key) => delete newBinding[key]);
+          return newBinding;
+        });
+        return {
+          recurse: true,
+          result: factory.createValues(
+            variables,
+            valueBindings,
+          ),
+        };
+      }
+      return {
+        recurse: false,
+        result: op,
+      };
+    },
+    expression: (op: Algebra.Expression, factory: Factory) => {
+      if (op.expressionType === 'term') {
+        // Materialize a term expression
+        return {
+          recurse: false,
+          result: factory.createTermExpression(materializeTerm(op.term, bindings)),
+        }
+      }
+      if (op.expressionType === 'aggregate' && 'variable' in op && bindings.has(termToString(op.variable))) {
+        // Materialize a bound aggregate operation.
+        // If strictTargetVariables is true, we throw if the expression target variable is attempted to be bound.
+        // Otherwise, we ignore this operation.
+        if (strictTargetVariables) {
+          throw new Error(`Tried to bind ${termToString(op.variable)} in a ${op.aggregator} aggregate.`);
+        } else {
+          return {
+            recurse: true,
+            result: op,
+          }
```

Comment:
So in this case the variable will not be assigned the value of the bound variable (while in the case of an extend for example the variable does get bound to the new value). But I'll be honest I'm also not 100% sure what could break by doing this :D

Reply:
If I understand correctly, it doesn't make much sense to pre-bind a bound aggregate, so ignoring it feels like the safest way forward.